crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -93,10 +97,15 @@ impl<'db> UnionBuilder<'db> {
             // means we shouldn't add it. Otherwise, add a new `UnionElement::StringLiterals`
             // containing it.
             Type::StringLiteral(literal) => {
+                let mut too_large = false;
                 let mut found = false;
                 for element in &mut self.elements {
                     match element {
                         UnionElement::StringLiterals(literals) => {
+                            if literals.len() >= MAX_UNION_LITERALS {
+                                too_large = true;
+                                break;
```

Comment:
Couldn't we return here directly and get rid of the `too_large` flags?
```suggestion
                                let replace_with = KnownClass::Str.to_instance(self.db);
                                return self.add(replace_with);
```

Reply:
Yes indeed, thank you. I had tried this initially (but without the `replace_with` variable) and had borrow checker issues on `self`, which I initially mistakenly thought had to with the iteration over `&mut self.elements`, so I moved it out of the loop. Then I realized the borrow checker actually had a problem with accessing `self.db` inside a call to `self.add` (which takes ownership of `self`) and fixed that, but failed to realize I could move it back inside the loop.

https://github.com/astral-sh/ruff/pull/17429